Add support for managing branches, PRs and commits

josecordaz opened the pull request:
Changes:
- implement create branches, pullrequests and commits
- implement listing for commits

Fix: #72
Fix: #73

stefanprodan commented on github/client_repository_commit.go:
```diff
@@ -0,0 +1,126 @@
+	}
+
+	return newCommit(c, nCommit),nil
+}
```
New line here please

stefanprodan commented on github/githubclient.go:
```diff
@@ -361,3 +392,20 @@ func (c *githubClientImpl) RemoveTeam(ctx context.Context, orgName, repo, teamNa
 	return handleHTTPError(err)
 }
+
+//func (c *githubClientImpl) ListRepoCommits(ctx context.Context, username ,repo,branch string) ([]*github.RepositoryCommit, error) {
```
Please remove this

stefanprodan commented on github/util.go:
```diff
@@ -151,6 +151,9 @@ func handleHTTPError(err error) error {
 func allPages(opts *github.ListOptions, fn func() (*github.Response, error)) error {
 	for {
+
+		fmt.Printf("%+v\n",opts)
```
Please don’t write to stdout from packages

stefanprodan commented on gitlab/client_repository_pullrequest.go:
```diff
@@ -0,0 +1,57 @@
+
+	bts ,_ := json.Marshal(prOpts)
+
+	fmt.Println(string(bts))
```
Please don’t write to stdout from packages

stefanprodan commented on gitprovider/client.go:
```diff
@@ -206,3 +206,27 @@ type DeployKeyClient interface {
+type PullRequestClient interface {
+	// Create creates a pull request with the given specifications.
+	Create(ctx context.Context,title,branch,baseBranch,description string)  error
+}
```
New line here

stefanprodan reviewed (changes requested):
@josecordaz please add tests for all the new functions and post the test results in here.

josecordaz commented:
GITLAB integration tests results:
```
GOROOT=/usr/local/Cellar/go/1.16.3/libexec #gosetup
GOPATH=/Users/joseordaz/go #gosetup
/usr/local/Cellar/go/1.16.3/libexec/bin/go test -c -o /private/var/folders/24/xpk2zmbj2552mxs2nhzsq8t00000gn/T/___9go_test_github_com_fluxcd_go_git_providers_gitlab.test -gcflags all=-N -l github.com/fluxcd/go-git-providers/gitlab #gosetup
/usr/local/Cellar/go/1.16.3/libexec/bin/go tool test2json -t /private/var/folders/24/xpk2zmbj2552mxs2nhzsq8t00000gn/T/dlvLauncher.sh /Applications/GoLand.app/Contents/plugins/go/lib/dlv/mac/dlv --listen=0.0.0.0:53144 --headless=true --api-version=2 --check-go-version=false --only-same-user=false exec /private/var/folders/24/xpk2zmbj2552mxs2nhzsq8t00000gn/T/___9go_test_github_com_fluxcd_go_git_providers_gitlab.test -- -test.v -test.paniconexit0 -test.v
=== RUN   Test_clientOptions_getTransportChain
--- PASS: Test_clientOptions_getTransportChain (0.00s)
=== RUN   Test_clientOptions_getTransportChain/all_roundtrippers
    --- PASS: Test_clientOptions_getTransportChain/all_roundtrippers (0.00s)
=== RUN   Test_clientOptions_getTransportChain/only_pre_+_auth
    --- PASS: Test_clientOptions_getTransportChain/only_pre_+_auth (0.00s)
=== RUN   Test_clientOptions_getTransportChain/only_cache_+_auth
    --- PASS: Test_clientOptions_getTransportChain/only_cache_+_auth (0.00s)
=== RUN   Test_makeOptions
--- PASS: Test_makeOptions (0.00s)
=== RUN   Test_makeOptions/no_options
    --- PASS: Test_makeOptions/no_options (0.00s)
=== RUN   Test_makeOptions/WithDomain
    --- PASS: Test_makeOptions/WithDomain (0.00s)
=== RUN   Test_makeOptions/WithDomain,_empty
    --- PASS: Test_makeOptions/WithDomain,_empty (0.00s)
=== RUN   Test_makeOptions/WithDestructiveAPICalls
    --- PASS: Test_makeOptions/WithDestructiveAPICalls (0.00s)
=== RUN   Test_makeOptions/WithPreChainTransportHook
    --- PASS: Test_makeOptions/WithPreChainTransportHook (0.00s)
=== RUN   Test_makeOptions/WithPreChainTransportHook,_nil
    --- PASS: Test_makeOptions/WithPreChainTransportHook,_nil (0.00s)
=== RUN   Test_makeOptions/WithPostChainTransportHook
    --- PASS: Test_makeOptions/WithPostChainTransportHook (0.00s)
=== RUN   Test_makeOptions/WithPostChainTransportHook,_nil
    --- PASS: Test_makeOptions/WithPostChainTransportHook,_nil (0.00s)
=== RUN   Test_makeOptions/WithOAuth2Token
    --- PASS: Test_makeOptions/WithOAuth2Token (0.00s)
=== RUN   Test_makeOptions/WithOAuth2Token,_empty
    --- PASS: Test_makeOptions/WithOAuth2Token,_empty (0.00s)
=== RUN   Test_makeOptions/WithConditionalRequests
    --- PASS: Test_makeOptions/WithConditionalRequests (0.00s)
=== RUN   Test_makeOptions/WithConditionalRequests,_exclusive
    --- PASS: Test_makeOptions/WithConditionalRequests,_exclusive (0.00s)
=== RUN   Test_DomainVariations
--- PASS: Test_DomainVariations (0.00s)
=== RUN   Test_DomainVariations/gitlab.com_domain
    --- PASS: Test_DomainVariations/gitlab.com_domain (0.00s)
=== RUN   Test_DomainVariations/custom_domain_without_protocol
    --- PASS: Test_DomainVariations/custom_domain_without_protocol (0.00s)
=== RUN   Test_DomainVariations/custom_domain_with_https_protocol
    --- PASS: Test_DomainVariations/custom_domain_with_https_protocol (0.00s)
=== RUN   Test_DomainVariations/custom_domain_with_http_protocol
    --- PASS: Test_DomainVariations/custom_domain_with_http_protocol (0.00s)
=== RUN   TestProvider
Running Suite: GitLab Provider Suite
====================================
Random Seed: 1619647949
Will run 11 of 11 specs

• [SLOW TEST:6.728 seconds]
GitLab Provider
/Users/joseordaz/go/src/github.com/josecordaz/go-git-providers/gitlab/integration_test.go:184
  should list the available organizations the user has access to
  /Users/joseordaz/go/src/github.com/josecordaz/go-git-providers/gitlab/integration_test.go:271
------------------------------
•
------------------------------
• [SLOW TEST:10.071 seconds]
GitLab Provider
/Users/joseordaz/go/src/github.com/josecordaz/go-git-providers/gitlab/integration_test.go:184
  should be possible to create a group project
  /Users/joseordaz/go/src/github.com/josecordaz/go-git-providers/gitlab/integration_test.go:328
------------------------------
•
------------------------------
• [SLOW TEST:6.963 seconds]
GitLab Provider
/Users/joseordaz/go/src/github.com/josecordaz/go-git-providers/gitlab/integration_test.go:184
  should update if the org repo already exists when reconciling
  /Users/joseordaz/go/src/github.com/josecordaz/go-git-providers/gitlab/integration_test.go:374
------------------------------
• [SLOW TEST:100.244 seconds]
GitLab Provider
/Users/joseordaz/go/src/github.com/josecordaz/go-git-providers/gitlab/integration_test.go:184
  should update teams with access and permissions when reconciling
  /Users/joseordaz/go/src/github.com/josecordaz/go-git-providers/gitlab/integration_test.go:417
------------------------------
•
------------------------------
• [SLOW TEST:9.018 seconds]
GitLab Provider
/Users/joseordaz/go/src/github.com/josecordaz/go-git-providers/gitlab/integration_test.go:184
  should be possible to create a user project
  /Users/joseordaz/go/src/github.com/josecordaz/go-git-providers/gitlab/integration_test.go:592
------------------------------
•••
Ran 11 of 11 Specs in 144.956 seconds
SUCCESS! -- 11 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestProvider (144.96s)
=== RUN   Test_getGitProviderPermission
--- PASS: Test_getGitProviderPermission (0.00s)
=== RUN   Test_getGitProviderPermission/pull
    --- PASS: Test_getGitProviderPermission/pull (0.00s)
=== RUN   Test_getGitProviderPermission/push
    --- PASS: Test_getGitProviderPermission/push (0.00s)
=== RUN   Test_getGitProviderPermission/admin
    --- PASS: Test_getGitProviderPermission/admin (0.00s)
=== RUN   Test_getGitProviderPermission/false_data
    --- PASS: Test_getGitProviderPermission/false_data (0.00s)
=== RUN   Test_getGitlabPermission
--- PASS: Test_getGitlabPermission (0.00s)
=== RUN   Test_getGitlabPermission/pull
    --- PASS: Test_getGitlabPermission/pull (0.00s)
=== RUN   Test_getGitlabPermission/push
    --- PASS: Test_getGitlabPermission/push (0.00s)
=== RUN   Test_getGitlabPermission/admin
    --- PASS: Test_getGitlabPermission/admin (0.00s)
=== RUN   Test_validateAPIObject
--- PASS: Test_validateAPIObject (0.00s)
=== RUN   Test_validateAPIObject/no_error_=>_nil
    --- PASS: Test_validateAPIObject/no_error_=>_nil (0.00s)
=== RUN   Test_validateAPIObject/one_error_=>_MultiError_&_InvalidServerData
    --- PASS: Test_validateAPIObject/one_error_=>_MultiError_&_InvalidServerData (0.00s)
=== RUN   Test_allGroupPages
--- PASS: Test_allGroupPages (0.00s)
=== RUN   Test_allGroupPages/one_page_only,_no_error
    --- PASS: Test_allGroupPages/one_page_only,_no_error (0.00s)
=== RUN   Test_allGroupPages/two_pages,_no_error
    --- PASS: Test_allGroupPages/two_pages,_no_error (0.00s)
=== RUN   Test_allGroupPages/four_pages,_error_at_second
    --- PASS: Test_allGroupPages/four_pages,_error_at_second (0.00s)
=== RUN   TestExampleOrganizationsClient_Get
Name: fluxcd-testing-public. Location: fluxcd-testing-public.--- PASS: TestExampleOrganizationsClient_Get (0.38s)
PASS

Debugger finished with the exit code 0
=== RUN   ExampleOrgRepositoriesClient_Get
--- PASS: ExampleOrgRepositoriesClient_Get (0.52s)

```

josecordaz commented:
GITHUB Integration tests results:
```
GOROOT=/usr/local/Cellar/go/1.16.3/libexec #gosetup
GOPATH=/Users/joseordaz/go #gosetup
/usr/local/Cellar/go/1.16.3/libexec/bin/go test -c -o /private/var/folders/24/xpk2zmbj2552mxs2nhzsq8t00000gn/T/___3go_test_github_com_fluxcd_go_git_providers_github.test github.com/fluxcd/go-git-providers/github #gosetup
/usr/local/Cellar/go/1.16.3/libexec/bin/go tool test2json -t /private/var/folders/24/xpk2zmbj2552mxs2nhzsq8t00000gn/T/___3go_test_github_com_fluxcd_go_git_providers_github.test -test.v -test.paniconexit0
=== RUN   Test_clientOptions_getTransportChain
--- PASS: Test_clientOptions_getTransportChain (0.00s)
=== RUN   Test_clientOptions_getTransportChain/all_roundtrippers
    --- PASS: Test_clientOptions_getTransportChain/all_roundtrippers (0.00s)
=== RUN   Test_clientOptions_getTransportChain/only_pre_+_auth
    --- PASS: Test_clientOptions_getTransportChain/only_pre_+_auth (0.00s)
=== RUN   Test_clientOptions_getTransportChain/only_cache_+_auth
    --- PASS: Test_clientOptions_getTransportChain/only_cache_+_auth (0.00s)
=== RUN   Test_makeOptions
--- PASS: Test_makeOptions (0.00s)
=== RUN   Test_makeOptions/no_options
    --- PASS: Test_makeOptions/no_options (0.00s)
=== RUN   Test_makeOptions/WithDomain
    --- PASS: Test_makeOptions/WithDomain (0.00s)
=== RUN   Test_makeOptions/WithDomain,_empty
    --- PASS: Test_makeOptions/WithDomain,_empty (0.00s)
=== RUN   Test_makeOptions/WithDestructiveAPICalls
    --- PASS: Test_makeOptions/WithDestructiveAPICalls (0.00s)
=== RUN   Test_makeOptions/WithPreChainTransportHook
    --- PASS: Test_makeOptions/WithPreChainTransportHook (0.00s)
=== RUN   Test_makeOptions/WithPreChainTransportHook,_nil
    --- PASS: Test_makeOptions/WithPreChainTransportHook,_nil (0.00s)
=== RUN   Test_makeOptions/WithPostChainTransportHook
    --- PASS: Test_makeOptions/WithPostChainTransportHook (0.00s)
=== RUN   Test_makeOptions/WithPostChainTransportHook,_nil
    --- PASS: Test_makeOptions/WithPostChainTransportHook,_nil (0.00s)
=== RUN   Test_makeOptions/WithOAuth2Token
    --- PASS: Test_makeOptions/WithOAuth2Token (0.00s)
=== RUN   Test_makeOptions/WithOAuth2Token,_empty
    --- PASS: Test_makeOptions/WithOAuth2Token,_empty (0.00s)
=== RUN   Test_makeOptions/WithConditionalRequests
    --- PASS: Test_makeOptions/WithConditionalRequests (0.00s)
=== RUN   Test_makeOptions/WithConditionalRequests,_exclusive
    --- PASS: Test_makeOptions/WithConditionalRequests,_exclusive (0.00s)
=== RUN   TestProvider
Running Suite: GitHub Provider Suite
====================================
Random Seed: 1619649086
Will run 8 of 8 specs

••••••••
Ran 8 of 8 Specs in 10.157 seconds
SUCCESS! -- 8 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestProvider (10.16s)
=== RUN   Test_getPermissionFromMap
--- PASS: Test_getPermissionFromMap (0.00s)
=== RUN   Test_getPermissionFromMap/pull
    --- PASS: Test_getPermissionFromMap/pull (0.00s)
=== RUN   Test_getPermissionFromMap/push
    --- PASS: Test_getPermissionFromMap/push (0.00s)
=== RUN   Test_getPermissionFromMap/admin
    --- PASS: Test_getPermissionFromMap/admin (0.00s)
=== RUN   Test_getPermissionFromMap/none
    --- PASS: Test_getPermissionFromMap/none (0.00s)
=== RUN   Test_getPermissionFromMap/false_data
    --- PASS: Test_getPermissionFromMap/false_data (0.00s)
=== RUN   Test_getPermissionFromMap/not_all_specifed
    --- PASS: Test_getPermissionFromMap/not_all_specifed (0.00s)
=== RUN   Test_validateAPIObject
--- PASS: Test_validateAPIObject (0.00s)
=== RUN   Test_validateAPIObject/no_error_=>_nil
    --- PASS: Test_validateAPIObject/no_error_=>_nil (0.00s)
=== RUN   Test_validateAPIObject/one_error_=>_MultiError_&_InvalidServerData
    --- PASS: Test_validateAPIObject/one_error_=>_MultiError_&_InvalidServerData (0.00s)
=== RUN   Test_allPages
--- PASS: Test_allPages (0.00s)
=== RUN   Test_allPages/one_page_only,_no_error
    --- PASS: Test_allPages/one_page_only,_no_error (0.00s)
=== RUN   Test_allPages/two_pages,_no_error
    --- PASS: Test_allPages/two_pages,_no_error (0.00s)
=== RUN   Test_allPages/four_pages,_error_at_second
    --- PASS: Test_allPages/four_pages,_error_at_second (0.00s)
=== RUN   ExampleOrganizationsClient_Get
--- PASS: ExampleOrganizationsClient_Get (0.08s)
=== RUN   ExampleOrgRepositoriesClient_Get
--- PASS: ExampleOrgRepositoriesClient_Get (0.64s)
PASS

Process finished with the exit code 0

```

stefanprodan reviewed (approved):
LGTM

Thanks @josecordaz please squash your commits.

josecordaz commented:
> LGTM
> 
> Thanks @josecordaz please squash your commits.

@stefanprodan  I don't have permissions to merge. Is there a workaround ?

stefanprodan commented:
@josecordaz I will merge it once you squash your commits.

josecordaz commented:
> @josecordaz I will merge it once you squash your commits.

@stefanprodan done 🙂

stefanprodan commented:
@josecordaz can you please read how commits should be made here: https://github.com/fluxcd/go-git-providers/blob/master/CONTRIBUTING.md#format-of-the-commit-message

From the docs:
- Limit the subject to 50 characters and write as the continuation of the sentence "If applied, this commit will ..."
- Explain what and why in the body, if more than a trivial change; wrap it at 72 characters.

Please reword your commit message and body, right now it looks like this:
```
added support to create branches, pullrequests and commits. Added lis… …

…tPage for commits. (github,gitlab)

Signed-off-by: Jose C Ordaz <josecordaz@gmail.com>

removed idea folder

Signed-off-by: Jose C Ordaz <josecordaz@gmail.com>

removed idea folder

Signed-off-by: Jose C Ordaz <josecordaz@gmail.com>

Fixed format, removed debug code

Signed-off-by: Jose C Ordaz <josecordaz@gmail.com>

Added integration tests, fixed stability to test gitlab integration tests

Signed-off-by: Jose C Ordaz <josecordaz@gmail.com>
```

josecordaz commented:
> @josecordaz can you please read how commits should be made here: https://github.com/fluxcd/go-git-providers/blob/master/CONTRIBUTING.md#format-of-the-commit-message
> 
> From the docs:
> 
>     * Limit the subject to 50 characters and write as the continuation of the sentence "If applied, this commit will ..."
> 
>     * Explain what and why in the body, if more than a trivial change; wrap it at 72 characters.
> 
> 
> Please reword your commit message and body, right now it looks like this:
> 
> ```
> added support to create branches, pullrequests and commits. Added lis… …
> 
> …tPage for commits. (github,gitlab)
> 
> Signed-off-by: Jose C Ordaz <josecordaz@gmail.com>
> 
> removed idea folder
> 
> Signed-off-by: Jose C Ordaz <josecordaz@gmail.com>
> 
> removed idea folder
> 
> Signed-off-by: Jose C Ordaz <josecordaz@gmail.com>
> 
> Fixed format, removed debug code
> 
> Signed-off-by: Jose C Ordaz <josecordaz@gmail.com>
> 
> Added integration tests, fixed stability to test gitlab integration tests
> 
> Signed-off-by: Jose C Ordaz <josecordaz@gmail.com>
> ```

@stefanprodan done

stefanprodan commented:
Signoff please: git commit --amend --no-edit --signoff
